### PR TITLE
Messages on init.

### PR DESCRIPTION
**Pull request by phlptp**
### Summary

If merged this pull request will fix a reported bug in received message during the init phase.  The iteration worked but no messages were available.  This PR adds a test case and fixes the bug.  

This PR ended up being a major refactor to resolve the filter timing issues for certain cases.  There are no major API changes but the underlying timing code has changed significantly.

### CODE REVIEW

**Review by eberleim — APPROVED** [review]

File: tests/helics/application_api/FilterAdditionalTests.cpp [hunk 1]
```diff
@@ -54,16 +54,23 @@ TEST_P(filter_type_tests, message_reroute_filter_object1)
 
     mFed->requestTimeAsync(1.0);
     fFed->requestTime(1.0);
+    // std::this_thread::sleep_for(std::chrono::milliseconds(2000));
+    // auto res = broker->query("root", "global_time_debugging");
+    // std::this_thread::sleep_for(std::chrono::milliseconds(2000));
+    // std::this_thread::yield();
+    // auto res2 = broker->query("root", "global_time_debugging");
```
Comment (nightlark) [inline tests/helics/application_api/FilterAdditionalTests.cpp:61]: Maybe something else to remove when not needed, along with a few other similar comments in other test files.

File: tests/helics/test_files/example_filters.json [hunk 2]
```diff
@@ -10,7 +10,7 @@
   "source_only": false, //indicator that the federate is only a source and is not expected to receive anything
   "uninterruptible": false, //indicator that the federate should only return requested times
   "coretype": "test", //the type of the core "test","zmq","udp","ipc","tcp","mpi"
-  "corename": "the name of the core", //this matters most for ipc and test cores, can be empty
+  "corename": "the name of the core1", //this matters most for ipc and test cores, can be empty
```
Comment (nightlark) [inline tests/helics/test_files/example_filters.json:13]:
```suggestion
  "corename": "the name of the core json", //this matters most for ipc and test cores, can be empty
```

File: src/helics/core/FilterFederate.cpp [hunk 3]
```diff
@@ -0,0 +1,995 @@
...
+                                                 dest,
+                                             handle,
+                                             key,
+                                             type_in,
+                                             type_out,
+                                             false);
+
+    auto cid = filt->core_id;
+    auto* retTarget = filt.get();
+    // auto actualKey = key;
```
Comment (nightlark) [inline src/helics/core/FilterFederate.cpp:604]: Comments to clean up later?

File: src/helics/core/FilterFederate.cpp [hunk 4]
```diff
@@ -0,0 +1,995 @@
...
+                    cmd.setAction(destFilter ? CMD_DEST_FILTER_RESULT : CMD_FILTER_RESULT);
+
+                    cmd.source_handle = FiltI->handle;
+                    cmd.source_id = mFedID;
+                    mDeliverMessage(cmd);
+                }
+            }
+        } else {
+            assert(false);
+            // this is an odd condition (not sure what to do yet)
```
Comment (nightlark) [inline src/helics/core/FilterFederate.cpp:145]: Is this something that should be a TODO?

Is there a way to recover or gracefully shutdown so assert isn't needed?

**Review by nightlark — APPROVED** [review]

File: src/helics/core/FilterFederate.cpp (hunk 4 above)
Comment (phlptp) [reply src/helics/core/FilterFederate.cpp:145]: A lot of the FilterFederate code is pulled from the CommonCore so this just got copied over.  I don't think we have reached that condition so I haven't dealt with it yet.

File: src/helics/core/BrokerBase.cpp [hunk 5]
```diff
@@ -480,6 +480,12 @@ void BrokerBase::setLogLevels(int32_t consoleLevel, int32_t fileLevel)
 
 void BrokerBase::addActionMessage(const ActionMessage& m)
 {
+    // if (m.dest_id == global_federate_id(131074) && m.source_id == global_federate_id(1879048194))
+    // {
+    ///   if (m.action() == CMD_TIME_REQUEST && m.actionTime>timeZero) {
+    //       printf("adding action message\n");
+    //  }
+    //}
```
[the remark this answers is not in the document]
Comment (phlptp) [reply src/helics/core/BrokerBase.cpp:488]: I am going to leave this in for a little while yet,  it won't be in the release code

**Pull request merged**